Delayed and Concurrent Job Executor, take 3

matbryan52 opened the pull request:
Continues #1158 with a DelayedJobExecutor supporting chunked results and an integrated method to perform partition-level merges using Dask functions.

Closes #1146, closes #922, closes #224, closes #2

Test cases cover:
- One or multiple UDFs
- With or without ROI
- Complex and simple merge functions (StdDevUDF, notably)
- Nav, Sig, Single buffers (including PickUDF)

The commit history starts from @uellue 's branch but has been rebased to have unrelated commits removed. As a result the ConcurrentJobExecutor is contained within this PR, though I haven't worked on/looked much at it.

## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [N/A] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

sk1p commented on src/libertem/common/buffers.py:
```diff
@@ -561,6 +570,28 @@ def __repr__(self):
             self._kind, self._dtype, self._extra_shape
         )
 
+    def update_data(self, data: nt.ArrayLike, force: bool = False) -> None:
```
```suggestion
    def update_data(self, data: "nt.ArrayLike", force: bool = False) -> None:
```

Using a string here delays the evaluation to type checking time.

matbryan52 replied:
I've implemented this change, thanks!

uellue commented:
Regarding the test failure in 3.9: https://dev.azure.com/LiberTEM/LiberTEM/_build/results?buildId=1481&view=logs&j=2c8ac9e1-3490-503e-dca6-5c76eb9f7e45&t=507d1be6-2f3f-540e-2620-1c2dec09a2b4&l=237

The second buffer is float32. Might be a numerical issue? Why is the dtype different, actually?

matbryan52 commented:
> Regarding the test failure in 3.9: https://dev.azure.com/LiberTEM/LiberTEM/_build/results?buildId=1481&view=logs&j=2c8ac9e1-3490-503e-dca6-5c76eb9f7e45&t=507d1be6-2f3f-540e-2620-1c2dec09a2b4&l=237
> 
> The second buffer is float32. Might be a numerical issue? Why is the dtype different, actually?

Yes you're probably right that it's numerical. It's comparing the result of `StdDevUDF` to `np.std` on the data itself. I set an absolute tolerance in `np.allclose` as 1e-3 but maybe that's still too tight! Given that the data is random I can see why this might sometimes fail. Could either make it deterministic or make the tolerance looser.

uellue commented:
> I set an absolute tolerance in `np.allclose` as 1e-3 but maybe that's still too tight!

Ah, I see it is the reference data that is float32? You can use `np.std(..., dtype=np.float64)` to do the reference with high precision.

matbryan52 commented:
> Ah, I see it is the reference data that is float32? You can use `np.std(..., dtype=np.float64)` to do the reference with high precision.

I've added the higher precision compute, still possible to make it deterministic though, if that would help.

uellue commented:
> I've added the higher precision compute, still possible to make it deterministic though, if that would help.

Actually, it is good if it is non-deterministic because that's how we catch such borderline cases. :-)

uellue commented:
About the `dask_merge()`: I was wondering about three items here:

* Could this be useful for other purposes as well, so that it would make sense to not limit this to Dask arrays?

* I have a somewhat uneasy gut feeling with having two independent functions in an UDF that should produce the same result along different pathways. It could add more ways to make a mistake, i.e. more maintenance and test burden. On the other hand it is the only way I can see to properly "package" this functionality with a UDF.

* With Dask arrays now supporting assignment, does this still constitute a big advantage over the "traditional" merge? See the following example with assignments aligned with chunk boundaries:

![image](https://user-images.githubusercontent.com/25689052/146175758-95fc14fe-32eb-4f40-85a3-b72827706d2d.png)

That task graph looks pretty clean. That means consecutive assignments into a `kind='nav'` result buffer that was allocated with the correct chunks could work pretty well, i.e. we could just work with our default merge. The only thing to pay attention to would be to allocate the buffer with chunks that make sense. Even if that is not the case, splitting up the assignment over chunks should be a relatively cheap operation.

Maybe it is sth to benchmark first, and only if we see a strong difference between the assignment and the bits in `contrib.daskadapter` we introduce `dask_merge()`?

uellue commented:
I'll go ahead and write a benchmark. :-)

matbryan52 commented:
> About the `dask_merge()`: I was wondering about three items here:
> 
> * Could this be useful for other purposes as well, so that it would make sense to not limit this to Dask arrays?
> * I have a somewhat uneasy gut feeling with having two independent functions in an UDF that should produce the same result along different pathways. It could add more ways to make a mistake, i.e. more maintenance and test burden. On the other hand it is the only way I can see to properly "package" this functionality with a UDF.
> * With Dask arrays now supporting assignment, does this still constitute a big advantage over the "traditional" merge? See the following example with assignments aligned with chunk boundaries:
> 
> That task graph looks pretty clean. That means consecutive assignments into a `kind='nav'` result buffer that was allocated with the correct chunks could work pretty well, i.e. we could just work with our default merge. The only thing to pay attention to would be to allocate the buffer with chunks that make sense. Even if that is not the case, splitting up the assignment over chunks should be a relatively cheap operation.
> 
> Maybe it is sth to benchmark first, and only if we see a strong difference between the assignment and the bits in `contrib.daskadapter` we introduce `dask_merge()`?

I agree that it's not ideal to have two pathways.

The current implementation can work via the normal routes if `dask_merge` is not defined, in this case it wraps the merge function in a Delayed and then puts the results into the final buffers via assignment as you say. The thing is that we have to do a copy within the Delayed merge wrapper in order to let the merge do its job, then copy this data into the final buffer after the call. In addition, it makes the merge process sequential (as we give the merge its part of the results buffer to work on for the duration), whereas using a dask_merge() can let the Dask scheduler compute all intermediate results and eventual post-processing before the once-and-for-all merge call.

It would be good to see a benchmark, though.

uellue commented:
A quick reproducer for scheduling issues on a LiberTEM Dask cluster:

```python
import numpy as np

from libertem.api import Context
from libertem.executor.delayed import DelayedJobExecutor
from libertem.udf.sum import SumUDF  # No CUDA support in this one


if __name__ == '__main__':
    # Make a LiberTEM Dask cluster and set it as default for Dask
    ctx = Context.make_with("dask-make-default")
    del_ctx = Context(executor=DelayedJobExecutor())

    ds = ctx.load('memory', data=np.zeros((32, 32, 64, 64)))

    # No problem since scheduled properly
    res_sum = ctx.run_udf(dataset=ds, udf=SumUDF())

    # Scheduled also on CUDA and service workers.
    # CUDA will throw an error
    del_res_sum = del_ctx.run_udf(dataset=ds, udf=SumUDF())
    del_res_sum['intensity'].raw_data.compute()
```

Result:

```
Exception: ValueError("UDF backends are ('numpy',), supported on CUDA are 'cuda' and 'cupy'")

Traceback (most recent call last):
  File "C:\Users\dwebe\Documents\src\LiberTEM\prototypes\delayed\reproducer.py", line 21, in <module>
    del_res_sum['intensity'].raw_data.compute()
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\dask\base.py", line 285, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\dask\base.py", line 567, in compute
    results = schedule(dsk, keys, **kwargs)
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\client.py", line 2672, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\client.py", line 1981, in gather
    return self.sync(
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\client.py", line 852, in sync
    return sync(
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\utils.py", line 354, in sync
    raise exc.with_traceback(tb)
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\utils.py", line 337, in f
    result[0] = yield future
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\tornado\gen.py", line 762, in run
    value = future.result()
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\distributed\client.py", line 1846, in _gather
    raise exception.with_traceback(traceback)
  File "C:\Users\dwebe\.conda\envs\libertem39\lib\site-packages\dask\utils.py", line 34, in apply
    return func(*args, **kwargs)
  File "c:\users\dwebe\documents\src\libertem\src\libertem\executor\delayed.py", line 267, in task_wrap
    res = task(*args, **kwargs)
  File "c:\users\dwebe\documents\src\libertem\src\libertem\udf\base.py", line 1433, in __call__
    return UDFRunner(udfs).run_for_partition(
  File "c:\users\dwebe\documents\src\libertem\src\libertem\udf\base.py", line 1716, in run_for_partition
    numpy_udfs, cupy_udfs = self._udf_lists(device_class)
  File "c:\users\dwebe\documents\src\libertem\src\libertem\udf\base.py", line 1692, in _udf_lists
    raise ValueError(f"UDF backends are {backends}, supported on CUDA are "
ValueError: UDF backends are ('numpy',), supported on CUDA are 'cuda' and 'cupy'
```

As far as I know, delayed objects can't have workers "baked" into them. I can see several options here:

* Just support CPU for now in the `DelayedJobExecutor`
* Figure out how to supply the required resources together with the delayed object(s)
* Figure out how to supply the required workers together with the delayed object(s)
* Require the user to specify workers or resources
* Make the cluster a bit more promiscuous and quietly allow CPU computations on GPU workers

uellue commented:
> * Figure out how to supply the required resources together with the delayed object(s)

Using `libertem.udf.base.get_resources_for_backends()` it was relatively straightforward to implement this:

```python
@classmethod
def get_resources(cls, udfs, user_backends=tuple()):
    if isinstance(udfs, UDF):
        udfs = [udfs]
    backends = [udf.get_backends() for udf in udfs]
    return get_resources_for_backends(backends, user_backends)
```

And use it like this:

```python
udf = MySumUDF()
resources = DelayedJobExecutor.get_resources(udf)
del_res_sum = del_ctx.run_udf(dataset=ds, udf=udf)
del_res_sum['intensity'].raw_data.compute(resources=resources)
```

Would that be OK as a public API?

matbryan52 commented:
> > * Figure out how to supply the required resources together with the delayed object(s)
> 
> Using `libertem.udf.base.get_resources_for_backends()` it was relatively straightforward to implement this:
> 
> ```python
> @classmethod
> def get_resources(cls, udfs, user_backends=tuple()):
>     if isinstance(udfs, UDF):
>         udfs = [udfs]
>     backends = [udf.get_backends() for udf in udfs]
>     return get_resources_for_backends(backends, user_backends)
> ```
> 
> And use it like this:
> 
> ```python
> udf = MySumUDF()
> resources = DelayedJobExecutor.get_resources(udf)
> del_res_sum = del_ctx.run_udf(dataset=ds, udf=udf)
> del_res_sum['intensity'].raw_data.compute(resources=resources)
> ```
> 
> Would that be OK as a public API?

In the back of my head I had the idea to make an `executor.compute()` method which is passes directly through to `dask.compute()` in the default case, but can also traverse (potentially multiple) `UDFData` instances and `DaskBufferWrapper` instances to get the dask arrays to give to the real `compute`, then if necessary re-wrap the results. This could also be used to inject the resource information if needed (and practical), without the user having to figure out the API. This still leaves the user free to use the methods you described if they need. How does that sound?

uellue commented:
> In the back of my head I had the idea to make an `executor.compute()` method which is passes directly through to `dask.compute()` in the default case, but can also traverse (potentially multiple) `UDFData` instances and `DaskBufferWrapper` instances to get the dask arrays to give to the real `compute`, then if necessary re-wrap the results. This could also be used to inject the resource information if needed (and practical), without the user having to figure out the API. This still leaves the user free to use the methods you described if they need. How does that sound?

I was wondering how users will actually interact with Dask array UDF results. Generally, I'd guess that creating a Dask array as opposed to just calculating the result outright makes sense for results that are somewhat large, i.e. where out-of-core and/or distributed processing have benefits over plain NumPy arrays. I can see three applications where that could happen:

* Data conversion and extraction, such as corrections, coordinate transforms, benefitting from high-performance implementations in LiberTEM and flexibility as opposed to implementing their own methods based on plain Dask.
* Transparent consolidation of back-end code in an application to have the same implementation for Hyperspy and LiberTEM. That allows people to still call `compute()` on the result as they are used to.
* Chaining LiberTEM UDFs, i.e. feeding the output of one UDF into the next UDF via intermediate Dask arrays. That can also include using LiberTEM for simulations, i.e. UDFs that inflate input data rather than reduce it. The advantage over regular Dask arrays would be much more flexibility in inputs, outputs and implementation.

From my gut feeling, for all these applications people will probably extract the Dask arrays from the UDF results, do more things with that, and THEN call compute() on the results. That means we'd have to pass the relevant info to that `compute()` call. Having to provide specific arguments there is a bit cumbersome.

Maybe the `DelayedJobExecutor` can be made "dumb" by default and run CPU-only like on a regular Dask cluster so that `compute()` works by default, and require an argument to activate GPU + CPU processing?

matbryan52 commented:
> I was wondering how users will actually interact with Dask array UDF results. Generally, I'd guess that creating a Dask array as opposed to just calculating the result outright makes sense for results that are somewhat large, i.e. where out-of-core and/or distributed processing have benefits over plain NumPy arrays.

From a usability standpoint, I've added a method to the executor `unwrap_results(structure)` which converts any nested Python structure containing `UDFData` or `DaskBufferWrapper` descendants into the same Python structure but using dicts and Dask arrays, respectively. The output can be passed to `dask.compute` and it knows how to traverse the structure and give you results in the same order, alternatively the user can unpack the arrays directly and do what they want with them.

I think this could actually be made into a LiberTEM util function for any structure with BufferWrappers to get the numpy results out (sometime I find returning a dict-of-BufferWrappers as results a bit cumbersome!).

> From my gut feeling, for all these applications people will probably extract the Dask arrays from the UDF results, do more things with that, and THEN call compute() on the results. That means we'd have to pass the relevant info to that `compute()` call. Having to provide specific arguments there is a bit cumbersome.

One option we have is to return the required resources information with the call to `run_udfs()`, or as a side-property so that a switched-on user can use it downstream themselves (potentially adding/accumulating as they go). I've added a property of the executor `executor.resources_if_available` which returns the required UDF resources for the last run.

I also implemented the `executor.compute()` method which I suggested, so we can consider it. This has a few features:

- It will automatically unwrap the results of `ctx.run_udf()` into dicts of Dask arrays
- It will pass resources to `dask.compute` in the following order of priority:
  1. directly from the user if the `resources` kwarg is set
  2. from the udfs passed by the user if provided
  3. from the udfs which were in the last run of `ctx.run_udf`
   4. None/Empty dict otherwise

> Maybe the `DelayedJobExecutor` can be made "dumb" by default and run CPU-only like on a regular Dask cluster so that `compute()` works by default, and require an argument to activate GPU + CPU processing?

I think this would be good default behaviour anyway, to avoid unexpected errors. I'll add a check to this effect.

matbryan52 commented on src/libertem/executor/delayed.py:
```diff
@@ -121,6 +121,13 @@ def get_available_workers(self):
         ])
 
+    @classmethod
+    def get_resources(cls, udfs, user_backends=tuple()):
```
This duplicates the function on line 171. Happy to use your and version/naming, though.

uellue replied:
Ah, didn't see it. Gone it is! :-)

uellue commented:
Benchmark results on ptycho:

```
-------------------------------------------------------------------------------------- benchmark 'echo': 3 tests ---------------------------------------------------------------------------------------
Name (time in ms)                  Min                   Max                  Mean             StdDev                Median                IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_echo_delayed_merge       492.5176 (1.0)        507.0606 (1.0)        499.6006 (1.0)       5.4820 (1.0)        500.4840 (1.0)       7.4330 (1.0)           2;0  2.0016 (1.0)           5           1
test_echo_delayed             830.7237 (1.69)       906.2164 (1.79)       881.8123 (1.77)     30.9945 (5.65)       889.5514 (1.78)     40.0247 (5.38)          1;0  1.1340 (0.57)          5           1
test_echo_baseline          6,816.6955 (13.84)    6,872.1159 (13.55)    6,846.8766 (13.70)    21.4435 (3.91)     6,847.0881 (13.68)    31.6079 (4.25)          2;0  0.1461 (0.07)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'sum': 3 tests ----------------------------------------------------------------------------------
Name (time in ms)               Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_sum_baseline          111.7458 (1.0)      118.7805 (1.0)      113.9792 (1.0)       2.3731 (1.0)      113.2749 (1.0)       2.7987 (1.0)           2;0  8.7735 (1.0)           9           1
test_sum_delayed_merge     229.2539 (2.05)     349.4252 (2.94)     280.5030 (2.46)     50.3463 (21.22)    294.0116 (2.60)     78.5746 (28.07)         2;0  3.5650 (0.41)          5           1
test_sum_delayed           530.2578 (4.75)     612.1997 (5.15)     559.8482 (4.91)     39.0515 (16.46)    533.8617 (4.71)     65.5267 (23.41)         1;0  1.7862 (0.20)          5           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------- benchmark 'sumsig': 3 tests -----------------------------------------------------------------------------------
Name (time in ms)                  Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_sumsig_baseline          108.1646 (1.0)      113.6919 (1.0)      110.9064 (1.0)       1.8807 (1.0)      111.0232 (1.0)       2.2934 (1.0)           4;0  9.0166 (1.0)          10           1
test_sumsig_delayed_merge     213.9514 (1.98)     279.4278 (2.46)     241.0153 (2.17)     34.5315 (18.36)    216.7537 (1.95)     62.4770 (27.24)         2;0  4.1491 (0.46)          5           1
test_sumsig_delayed           423.4515 (3.91)     522.1981 (4.59)     467.0225 (4.21)     40.5295 (21.55)    478.5244 (4.31)     61.8636 (26.97)         2;0  2.1412 (0.24)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

That means the special merge function gives us about 2x speed-up on that machine. Also note how efficient it is to do downstream calculations with large UDF-generated Dask arrays! It is not as fast as doing it inline, but still quite competitive. I'd guess that L3 cache efficiency for tiled processing in the direct path makes the difference here.

matbryan52 commented:
> Benchmark results on ptycho:
> 
> ```
> -------------------------------------------------------------------------------------- benchmark 'echo': 3 tests ---------------------------------------------------------------------------------------
> Name (time in ms)                  Min                   Max                  Mean             StdDev                Median                IQR            Outliers     OPS            Rounds  Iterations
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> test_echo_delayed_merge       492.5176 (1.0)        507.0606 (1.0)        499.6006 (1.0)       5.4820 (1.0)        500.4840 (1.0)       7.4330 (1.0)           2;0  2.0016 (1.0)           5           1
> test_echo_delayed             830.7237 (1.69)       906.2164 (1.79)       881.8123 (1.77)     30.9945 (5.65)       889.5514 (1.78)     40.0247 (5.38)          1;0  1.1340 (0.57)          5           1
> test_echo_baseline          6,816.6955 (13.84)    6,872.1159 (13.55)    6,846.8766 (13.70)    21.4435 (3.91)     6,847.0881 (13.68)    31.6079 (4.25)          2;0  0.1461 (0.07)          5           1
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
> ----------------------------------------------------------------------------------- benchmark 'sum': 3 tests ----------------------------------------------------------------------------------
> Name (time in ms)               Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
> -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> test_sum_baseline          111.7458 (1.0)      118.7805 (1.0)      113.9792 (1.0)       2.3731 (1.0)      113.2749 (1.0)       2.7987 (1.0)           2;0  8.7735 (1.0)           9           1
> test_sum_delayed_merge     229.2539 (2.05)     349.4252 (2.94)     280.5030 (2.46)     50.3463 (21.22)    294.0116 (2.60)     78.5746 (28.07)         2;0  3.5650 (0.41)          5           1
> test_sum_delayed           530.2578 (4.75)     612.1997 (5.15)     559.8482 (4.91)     39.0515 (16.46)    533.8617 (4.71)     65.5267 (23.41)         1;0  1.7862 (0.20)          5           1
> -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
> ---------------------------------------------------------------------------------- benchmark 'sumsig': 3 tests -----------------------------------------------------------------------------------
> Name (time in ms)                  Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> test_sumsig_baseline          108.1646 (1.0)      113.6919 (1.0)      110.9064 (1.0)       1.8807 (1.0)      111.0232 (1.0)       2.2934 (1.0)           4;0  9.0166 (1.0)          10           1
> test_sumsig_delayed_merge     213.9514 (1.98)     279.4278 (2.46)     241.0153 (2.17)     34.5315 (18.36)    216.7537 (1.95)     62.4770 (27.24)         2;0  4.1491 (0.46)          5           1
> test_sumsig_delayed           423.4515 (3.91)     522.1981 (4.59)     467.0225 (4.21)     40.5295 (21.55)    478.5244 (4.31)     61.8636 (26.97)         2;0  2.1412 (0.24)          5           1
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ```
> 
> That means the special merge function gives us about 2x speed-up on that machine. Also note how efficient it is to do downstream calculations with large UDF-generated Dask arrays! It is not as fast as doing it inline, but still quite competitive. I'd guess that L3 cache efficiency for tiled processing in the direct path makes the difference here.

Great results!

Any chance you can add StdDevUDF to your benchmark ? The `dask_merge` implementation is in the file `test/executor/test_delayed.py`.

uellue commented:
New results:

```
--------------------------------------------------------------------------------------- benchmark 'large': 3 tests --------------------------------------------------------------------------------------
Name (time in ms)                   Min                   Max                  Mean             StdDev                Median                IQR            Outliers     OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_large_delayed_merge       497.4620 (1.0)        505.2565 (1.0)        500.8654 (1.0)       3.0256 (1.0)        500.0528 (1.0)       4.3385 (1.0)           2;0  1.9965 (1.0)           5           1
test_large_delayed             883.2547 (1.78)       910.7725 (1.80)       898.1595 (1.79)     12.0298 (3.98)       898.3527 (1.80)     21.7054 (5.00)          2;0  1.1134 (0.56)          5           1
test_large_baseline          6,770.3100 (13.61)    6,843.2811 (13.54)    6,807.9222 (13.59)    28.9151 (9.56)     6,816.4793 (13.63)    43.9254 (10.12)         2;0  0.1469 (0.07)          5           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

--------------------------------------------------------------------------------------- benchmark 'std': 3 tests --------------------------------------------------------------------------------------
Name (time in ms)               Min                   Max                  Mean              StdDev                Median                 IQR            Outliers     OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_std_baseline          637.1177 (1.0)        646.2906 (1.0)        640.3563 (1.0)        3.7754 (1.0)        639.3520 (1.0)        5.4522 (1.0)           1;0  1.5616 (1.0)           5           1
test_std_delayed_merge     883.2280 (1.39)       997.6045 (1.54)       917.1209 (1.43)      49.1688 (13.02)      890.3565 (1.39)      64.2657 (11.79)         1;0  1.0904 (0.70)          5           1
test_std_delayed           999.5253 (1.57)     1,340.5703 (2.07)     1,186.6475 (1.85)     139.5470 (36.96)    1,180.7805 (1.85)     229.0933 (42.02)         2;0  0.8427 (0.54)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'sum': 3 tests ----------------------------------------------------------------------------------
Name (time in ms)               Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_sum_baseline          109.5887 (1.0)      119.7171 (1.0)      112.4367 (1.0)       3.6749 (1.0)      110.5399 (1.0)       3.2779 (1.0)           2;2  8.8939 (1.0)          10           1
test_sum_delayed_merge     226.3882 (2.07)     344.1458 (2.87)     271.9702 (2.42)     48.7391 (13.26)    277.7806 (2.51)     72.6046 (22.15)         1;0  3.6769 (0.41)          5           1
test_sum_delayed           521.3182 (4.76)     638.4355 (5.33)     578.9842 (5.15)     41.4770 (11.29)    577.9953 (5.23)     34.0500 (10.39)         2;0  1.7272 (0.19)          5           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------- benchmark 'sumsig': 3 tests -----------------------------------------------------------------------------------
Name (time in ms)                  Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_sumsig_baseline          108.0212 (1.0)      111.7881 (1.0)      110.3604 (1.0)       1.4957 (1.0)      110.8081 (1.0)       2.7380 (1.0)           2;0  9.0612 (1.0)          10           1
test_sumsig_delayed_merge     217.6294 (2.01)     332.8640 (2.98)     285.0837 (2.58)     47.8769 (32.01)    273.1704 (2.47)     72.4760 (26.47)         1;0  3.5077 (0.39)          5           1
test_sumsig_delayed           413.3366 (3.83)     526.9932 (4.71)     483.1794 (4.38)     46.2671 (30.93)    475.9123 (4.29)     64.5047 (23.56)         1;0  2.0696 (0.23)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

sk1p commented:
/azp run libertem.libertem-data

uellue commented:
A quick test if we even need `libertem.contrib.daskadapter.make_dask_array` now that we could just use an UDF to echo the dataset shows that the direct route is very fast in comparison, even competitive against plain LiberTEM:

```
--------------------------------------------- benchmark 'large': 1 tests ---------------------------------------------
Name (time in ms)            Min       Max      Mean   StdDev    Median      IQR  Outliers     OPS  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------
test_large_dsdirect     119.5704  170.8547  131.3488  18.1680  123.4137  14.7847       1;1  7.6133       8           1
----------------------------------------------------------------------------------------------------------------------
```

sk1p commented on benchmarks/executor/test_delayed.py:
```diff
@@ -80,6 +91,10 @@ def get_result_buffers(self):
             ),
         }
 
+    # Make sure we don't have default merge so that there is no default dask_merge
```
Do we need to decouple the "can use default merge" and "can use dask_merge" cases? Can we? Do we need to document the implication that implementing a `dask_merge` means you also need to implement a normal `merge`?

uellue replied:
I'm working on that switching logic at the moment. Good point that one might only need a `merge()` _or_ a `merge_all()` to do the job for both cases. I'll see if I can make sth work. The point here was that for benchmarking the back-end I wanted to make sure to have a defined setup, since we'll likely implement `merge_all()` for many UDFs

uellue replied:
Running UDFs that only do `merge_all()` with normal executors could possibly be implemented, but that can maybe be left for later if a real need become apparent.

However, if a UDF is only being used with the DelayedJobExecutor, one doesn't have to implement `merge()`: In the default nav merge case it can use the default implementation for both variants, i.e. nothing to implement. For other cases one can only override `merge_all()`, and running with a normal executor will raise a `NotImplementedError` from the default merge.

sk1p commented on src/libertem/udf/base.py:
```diff
@@ -33,6 +33,7 @@
 from libertem.executor.base import Environment, JobExecutor
 
 if TYPE_CHECKING:
+    from typing import OrderedDict
```
Btw. for using many types that are added in a newer Python version, we also depend on the "polyfill" `typing_extensions`, which is especially needed if you want to actually instantiate such a type using the same import (vs. having to import the corresponding type from `collections` in addition, for example)

uellue replied:
I tried, but didn't get that to work any better than it is now. I hope that's ok!

uellue replied:
That threw an error when I tested with Python 3.6, unfortunately. But I am not sure if I did sth wrong, maybe.

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
Actually, we could consider writing a JOSS paper about the integration at some point. Combining Dask arrays and LiberTEM give fundamentally new and pretty unique capabilities.

sk1p commented on src/libertem/udf/base.py:
```diff
@@ -1777,54 +1935,71 @@ def run_for_dataset_sync(
+                part_results=part_results,
+                task=task
+            )
+            # Explicit indexing for compatibility with Dask.delayed
+            self._udfs = res[0]
+            damage = res[1]
```
Are these three lines really still necessary? If so, please remind me why, and we probably should add a test case that demonstrates this

uellue replied:
Yes, not necessary anymore thanks to @matbryan52 's work! :-)

sk1p commented on tests/executor/test_functional.py:
```diff
@@ -0,0 +1,279 @@
+    for kwargs in load_kwargs:
+        ds = ctx.load(**kwargs)
+        udfs = _make_udfs(ds)
+        roi = None
```
```suggestion
```

sk1p reviewed (changes requested):
Thanks a lot @matbryan52 and @uellue for taking charge of this issue :) I think this is almost ready to be merged, my comments are mostly about smaller details.

One thing to consider is that people may have older Python virtualenvs which already have dask installed, which may be too old for the assignment-style we are using here. I think we should bump our minimum version on Python 3.7+ to whatever is the version that introduced this.

uellue commented on azure-pipelines.yml:
```diff
@@ -142,6 +142,8 @@ stages:
         displayName: 'Use Python 3.9'
         inputs:
           versionSpec: '3.9'
+      - bash: sudo apt install -y pandoc graphviz
```
I'm a bit 50/50 between adding a dependency and adding complexity to the build and test chain. Since `graphviz` seems OK as a dependency, I'd opt to just depend on it so that the code is always executed reliably when tests run.

sk1p replied:
Is it easy enough to install on Windows - i.e. does the installer .exe work as advertised, and also together with the python wrapper library?

uellue replied:
I have it running on my Win 10 system and don't recall any issues with the installer. The Python wrapper is also just `pip install ...`, and then calls the binary.

uellue replied:
Having `.visualize()` is generally nice when working with Dask arrays, so it shouldn't be too much of a burden for LiberTEM devs. :-)

sk1p replied:
Perfect, then I'm fine with depending on graphviz :+1:

uellue commented on src/libertem/executor/concurrent.py:
```diff
@@ -0,0 +1,166 @@
+
+class ConcurrentJobExecutor(JobExecutor):
+    '''
+    :class:`JobExecutor` that uses :mod:`python.concurrent.futures`.
```
It was a typo, should be ``:mod:`python:concurrent.futures` ``

uellue commented on src/libertem/executor/utils/delayed_unpack.py:
```diff
@@ -0,0 +1,132 @@
+        return f'{self.__class__.__name__}({self.cls}, {self.args}, {self.kwargs})'
+
+
+def flatten_nested(el, unpackable_types=None, ignore_types=None):
```
I tried, but failed to make mypy accept it. Without that I may very well add wrong type annotations. I'd rather leave it as-is. If you think it is important, maybe you could have a look?

uellue replied:
As a pragmatic approach since these are internal utility functions: What about adding type annotations at the time we want to make changes and think "It would be nice to have type annotations now"?

sk1p replied:
I can have a look at it. If we want to delay adding type annotations, I would prefer if we had docstrings for these. It's mostly about properly documenting these, so the next time they are touched (or code using them is touched!) you don't have to read and understand the whole function bodies, as often type annotations really clarify the details enough. A docstring would serve the same purpose, even though it is not machine understandable.

uellue replied:
@matbryan52 maybe you could add some basic docstrings, then? Both @sk1p and I would have to reverse-engineer this code.

sk1p replied:
Looks like this one is now resolved in 3731ee72f4dca5dfaf76fef0e2a054579e4f645a c8c0b0d4945fb74dae63a85bf80f33e53f16ba2a 22bccbce9e57ad22a25becd53c9607371949a9fe - thanks @matbryan52!

uellue commented on src/libertem/executor/delayed.py:
```diff
@@ -0,0 +1,417 @@
+            progress: bool = False, corrections: Optional[CorrectionSet] = None,
+            backends: Optional[BackendSpec] = None, dry: bool = False,
+            iterate: bool = True) -> Generator["UDFResults", None, None]:
+        if iterate:
+            raise NotImplementedError(
+                "Iterating over partial results is currently not supported "
+                "with the DelayedJobExecutor."
+            )
```
For this and the following comment for the async case I don't really understand the use case. It just doesn't make sense to me to iterate over delayed partial results or generate them asynchronously. Do you have any proposal? For that reason I'd rather leave it unimplemented until there's a clear application instead of adding sth to the public API that we'll have to support in the future and that may or may not make any sense. ALso, that saves us all the testing code for it as well. I'd prefer to add things only with a clear use case, not just for completeness sake

sk1p replied:
Why does this need changes to the public API? The `Context` already supports running in sync/async iter/non-iter combinations, for example:

```python
from libertem.api import Context
from libertem.executor.delayed import DelayedJobExecutor
from libertem.udf.sum import SumUDF

ctx = Context(executor=DelayedJobExecutor())
ds = ctx.load("auto", path="...")
res = await ctx.run_udf(dataset=ds, udf=SumUDF(), sync=False)  # errs our with the above `NotImplementedError`
for res in ctx.run_udf_iter(dataset=ds, udf=SumUDF()):
    pass  # similarly errs out
```

Currently, this means that `DelayedJobExecutor` cannot be used in places where other executors just work. Specifically, think about integration of LiberTEM into other software - if it is written in an `async`/`await`-style, it will probably also use the `async` API of LiberTEM. I would prefer if things just work, vs having to maintain a support matrix for executors.

uellue replied:
My memory is a bit fuzzy, but I recall that I realized "hey, we should at least test the full API", and it didn't work out of the box. We clearly communicate that the DelayedJobExecutor is experimental, for that reason I just raise `NotImplementedError` for the cases that are not well-supported yet.

Yes, I agree for all other executors that we should support everything. The `DelayedJobExecutor` is just an odd one in the mix for the special task of generating Dask arrays. It would break an embedding application anyway, resp. require special handling of the result.

uellue replied:
TL; DR I'd strongly prefer the `NotImplementedError` over additional work to support iteration and async at this time.

sk1p replied:
This is done in 4e8d168b - according to the test cases, it was enough to remove the exceptions thrown in `run_for_dataset_{a,}sync` (and it is only supported on Python 3.7).

uellue commented on src/libertem/executor/delayed.py:
```diff
@@ -0,0 +1,417 @@
+            corrections=corrections, backends=backends, dry=dry, iterate=iterate
+        )
+
+    def run_for_dataset_async(self, *args, **kwargs):
+        raise NotImplementedError(
+            "Async execution is currently not supported with the DelayedJobExecutor."
+        )
```
See https://github.com/LiberTEM/LiberTEM/pull/1170#discussion_r782992374

uellue commented on src/libertem/udf/base.py:
```diff
@@ -1777,54 +1935,71 @@ def run_for_dataset_sync(
+        corrections: Optional[CorrectionSet] = None,
+        backends: Optional[BackendSpec] = None,
+        dry: bool = False,
+        iterate: bool = True
```
It really depends on what the user implements in `UDF.get_results()` and what partition size we have. It can be quite significant and is not in our control, that's why I'd skip it where possible.

uellue commented on src/libertem/udf/base.py:
```diff
@@ -1413,6 +1525,31 @@ def __repr__(self):
+            udf.clear_views()
+        return UDFResults(
+            buffers=tuple(
+                # Explicit indexing for compatibility with Dask.delayed
```
Yes, I think it was a leftover from refactoring. :-)

uellue replied:
Also, we don't have to overwrite `self.udfs` and `damage` anymore thanks to @matbryan52 's code unpacking the results. I've removed that from the code.

uellue commented on tox.ini:
```diff
@@ -53,9 +53,10 @@ deps=
     mrcfile
     pyxem
     pims
+    scikit-image<0.19
```
Since it is an issue within `pyxem`, I'm hoping there will be a fix released at some time. Since it is only used for testing, we can maybe skip these tests for Python 3.10 until `pyxem` is fixed?

sk1p replied:
Sounds good - 3.10 support is still waiting on a few releases anyway (most importantly numba, but primesieve is also currently built from source on 3.10, which would be painful on Windows...)

sk1p replied:
See also https://github.com/LiberTEM/LiberTEM/issues/1133

sk1p commented on src/libertem/executor/utils/dask_buffer.py:
```diff
@@ -0,0 +1,156 @@
+from .dask_inplace import DaskInplaceWrapper
+
+
+class DaskBufferWrapper(BufferWrapper):
```
Someting I missed in my first review: the `BufferWrapper` supports `__array__`, so you can use `np.array(res['intensity'])` or similar code to get the result as a `numpy` array. This could be an avenue for writing code that is compatible with both `DelayedJobExecutor` and the others. Currently, this gives a "ValueError: object __array__ method not producing an array" - I think something like this could work:

```python
def __array__(self):
    return np.array(self.data)
```

Speaking of compatibility, and sorry this remark comes so late in the process: have you considered, instead of returning the dask array in `.data`, to keep `.data` eager, and adding a `.delayed` attribute instead? As far as I can see, this would mean results from the `DelayedJobExecutor` would be fully compatible with other executors, and only add the additional functionality as additional API. Is there anything in the grand design of things that speaks against this, other than having to rewrite a bunch of tests and examples (which I volunteer to help with)?

sk1p replied:
I also had a look at `dask.array`, and sadly there doesn't seem to be an equivalent `__dask_array__` method we can implement to support `da.array(res['intensity'])`-style conversion to a dask array...

uellue replied:
@sk1p good idea with the `.delayed`. How should we handle `.raw_data` in this case? Add a `.raw_delayed`?

sk1p replied:
:+1: - would also be happy with more verbose but clearly symmetric names of `.data_delayed` and `.raw_data_delayed` (but that's now a bit bike shedding :))

sk1p replied:
This is done in 9c833965 and follow-up commits

sk1p commented on src/libertem/web/jobs.py:
```diff
@@ -1,12 +1,15 @@
 import logging
```
I think this is an artifact from rebasing - a lot of this code was moved to `libertem.web.engine` for #1172, so the changes in this PR need to be ported to that code. I think this was mostly getting the `UDFRunner` subclass from the executor?

sk1p commented:
/azp run libertem.libertem-data

sk1p commented:
/azp run libertem.libertem-data

sk1p commented:
> Sry should have waited with merging the other PR, created a conflict unfortunately.

Should I take care of the conflict?

sk1p commented:
> Sry should have waited with merging the other PR, created a conflict unfortunately.

Conflict is now resolved

uellue commented:
@matbryan52 Thx for the docs update! Looks much clearer now. I'll do a complete read-through, then this should be ready to go, right?

matbryan52 commented:
> @matbryan52 Thx for the docs update! Looks much clearer now. I'll do a complete read-through, then this should be ready to go, right?

For me yes, I haven't checked through all the changes since I left before Christmas, though, but I am sure you and sk1p have it covered.

There is an open review point about moving minimum compatibility to py3.7, that seems like a larger question than just this PR, however.

uellue commented:
/azp run libertem.libertem-data

uellue commented:
Regarding https://github.com/LiberTEM/LiberTEM/pull/1170#issuecomment-1013256333, in the example I've "hacked" some internals to pretend not having PyFFTW. Maybe there should be a way to allow threaded executors to deactivate it "officially"?

sk1p commented:
> Regarding [#1170 (comment)](https://github.com/LiberTEM/LiberTEM/pull/1170#issuecomment-1013256333), in the example I've "hacked" some internals to pretend not having PyFFTW. Maybe there should be a way to allow threaded executors to deactivate it "officially"?

Hmm, I think PyFFTW should work in a threaded environment. It's possible that the cause for this issue is the PyFFTW cache - we can disable that to check if that's the case. Or the internal threading of PyFFTW - maybe also try to set the number of workers to 1.

If it turns out that PyFFTW is generally incompatible with threaded executors, I think #400 is the right fix - one could think about using the environment to convey information about what libraries are available in the given situation. Possibly the `xp` = `np`/`cp` selector could be hanlded similarly, instead of having the selection code somewhere in the `UDFRunner`.

Thoughts?

matbryan52 commented:
I think this could be tested by commenting this line: https://github.com/LiberTEM/LiberTEM-blobfinder/blob/master/src/libertem_blobfinder/base/correlation.py#L16 in lt-blobfinder to see if the cache is the source of the error ?

I don't have the example to hand to test this myself at this moment.

matbryan52 commented:
> I think this could be tested by commenting this line: https://github.com/LiberTEM/LiberTEM-blobfinder/blob/master/src/libertem_blobfinder/base/correlation.py#L16 in lt-blobfinder to see if the cache is the source of the error ? My understanding is that the execution will have extra overhead from PyFFT object creation but still re-use the FFT plan.
> 
> I don't have the example to hand to test this myself at this moment.

A bit more testing, firstly I was able to reproduce the problem in the hyperspy notebook on Ubuntu. Disabling the above line 16 for enabling the pyfftw cache didn't solve it, nor did forcing pyfftw threads to 1. The bug did go away by forcing Dask to use only a single thread, however.

Interestingly, I tried a MWE, just running the correlation UDF, for either a Memory, Raw and Dask dataset outside of a notebook, and I couldn't reproduce the NaNs yet.

uellue commented:
/azp run libertem.libertem-data

uellue commented on src/libertem/utils/threading.py:
```diff
@@ -138,15 +200,30 @@ def set_num_threads(n):
+                    with lock:
+                        __threadpool_counter -= 1
+                        if __threadpool_counter == 0 and __threadpool_limiter is not None:
+                            __threadpool_limiter.__exit__(None, None, None)
```
@sk1p Here I am not sure what to do. `None, None, None` would be only the "happy case". What should we do if the `finally:` comes after an exception?

sk1p replied:
Hmm. I guess this will work with the current implementation of `_ThreadpoolLimiter.__exit__`, which just ignores any exceptions. I also can't currently make up a reason why it would behave differently in the case of exceptions, but maybe I'm not thinking creatively enough :) Instead of using the context manager API of `_ThreadpoolLimiter`, can we explicitly call `restore_original_limits` instead? It seems to be a quasi-stable interface, with an original name `_ThreadpoolLimiter.unregister` being kept as a backwards-compatible alias.

Thoughts?

uellue replied:
Ok, sounds good! I'm testing it locally now.

sk1p replied:
As a justification for using the almost-internal API: at least we will see the breakage, which is not the case with the nesting approach...

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

uellue commented:
/azp run libertem.libertem-data

sk1p commented:
Other than the `threadpoolctl` discussion above, the changes for threading/pyfftw look good to me :+1: It also looks like, at least on Linux, the pin for pytest-asyncio works fine.

uellue commented:
/azp run libertem.libertem-data

sk1p commented:
It looks like the issues from my last review (and from the pyfftw, pytest-asyncio fallout) are now all tackled - is there anything from your perspective that still needs to be done before merging, @matbryan52  @uellue?

uellue commented:
> It looks like the issues from my last review (and from the pyfftw, pytest-asyncio fallout) are now all tackled - is there anything from your perspective that still needs to be done before merging, @matbryan52 @uellue?

Good to go from my perspective!

matbryan52 commented:
> Good to go from my perspective!

Mine too, let's go!

uellue commented:
🚀🚀🚀  🎉🎉🎉 Congrats to everyone! :-)
